mqtt-client/main.py

```diff
@@ -174,107 +513,235 @@ def input_status(self, status):
+        )()
+
+    @background_task
+    def background_task_total_energy(self):
+        self._create_background_task(
+            'energy',
+            self.webinterface.get_total_energy,
+            self._process_total_energy
+        )()
+
+    def _process_sensor_status(self, sensor_config, json_data):
+        mqtt_messages = []
+        data_list = list(filter(None, json_data.get('status', [])))
+        for sensor_id, sensor_value in enumerate(data_list):
+            sensor_data = {'id': sensor_id,
+                           'name': self._sensors.get(sensor_id).get('name'),
+                           'value': float(sensor_value) + float(self._sensors.get(sensor_id).get('offset')),
+                           'timestamp': self._timestamp2isoformat()}
+            mqtt_messages.append({'topic': sensor_config.get('topic').format(id=sensor_id),
+                                  'message': sensor_data})
+        return mqtt_messages
+
+    def _process_realtime_power(self, sensor_config, json_data):
+        mqtt_messages = []
+        json_data.pop('success')
+        for module_id, values in json_data.items():
+            for input_id, sensor_values in enumerate(values):
+                sensor_data = {'sensor_id': input_id,
+                               'module_id': module_id,
+                               'name': self._power_modules.get(int(module_id)).get(int(input_id)).get('name'),
+                               'voltage': sensor_values[0],
+                               'frequency': sensor_values[1],
+                               'current': sensor_values[2],
+                               'power': sensor_values[3],
+                               'timestamp': self._timestamp2isoformat()}
+                mqtt_messages.append({'topic': sensor_config.get('topic').format(module_id=module_id, sensor_id=input_id),
+                                      'message': sensor_data })
+        return mqtt_messages
+
+    def _process_total_energy(self, sensor_config, json_data):
+        mqtt_messages = []
+        json_data.pop('success')
+        for module_id, values in json_data.items():
+            for input_id, sensor_values in enumerate(values):
+                sensor_data = {'sensor_id': input_id,
+                               'module_id': module_id,
+                               'name': self._power_modules.get(int(module_id)).get(int(input_id)).get('name'),
+                               'day': sensor_values[0],
+                               'night': sensor_values[1],
+                               'timestamp': self._timestamp2isoformat()}
+            mqtt_messages.append({'topic': sensor_config.get('topic').format(module_id=module_id, sensor_id=input_id),
+                                  'message': sensor_data})
+        return mqtt_messages
+
+    def _create_background_task(self, sensor_type, data_retriever, data_processor):
+        def background_function():
+            if self._enabled:
```

Comment:
If the plugin is installed and running, it will be disabled since there's not yet configured. This means that these background tasks won't be running and the user would have to configure the plugin and then do a restart. I think it would be a good idea to have something like:

```
while not self._enabled:
    time.sleep(2)
sensor_config = self._sen...
...
```

Similar for the `while frequency >= 10`, it might be better to put the whole block in a `while True`, and then check the frequency and if it's less than 10, just sleep for a second and `continue`.

Generally speaking, for a background task, it's best to put everything in a `while True:`, and continue the loop with a short sleep as long as it's not correctly configured. This makes sure that as soon as configuration changes, the background task will be running and will pick up the change.

Reply:
Change made in [this commit](https://github.com/DaGolle/openmotics-plugins/commit/4beb03c463c2b71be96a8cccded481fa4611f001).